lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,293 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'avro/io'
+require 'json'
+
+module Avro
+  module IO
+    # JsonDecoder keeps track of the data and reads it using a given data type.
+    class JsonDecoder
+      # memo-ized map of Schema::PrimitiveSchema objects so we're not allocating an object for every validation
+      PRIMITIVE_SCHEMAS = Hash[Schema::PRIMITIVE_TYPES_SYM.map { |k| [k, Schema::PrimitiveSchema.new(k)] }]
+
+      attr_accessor :data
+
+      def io_reader(reader)
+        datum = reader.read
+        datum = '""' if datum.empty?
+        data = JSON.parse(datum, {:quirks_mode => true})
+        @data = data
+
+        self
+      end
+
+      def read_null
+        validate(:null, data)
+        data
+      end
+
+      def read_boolean
+        validate(:boolean, data)
+        data
+      end
+
+      def read_int
+        validate(:int, data)
+        Integer(data)
+      end
+
+      def read_long
+        validate(:long, data)
+        Integer(data)
+      end
+
+      def read_float
+        validate(:float, data)
+        Float(data)
+      end
+
+      def read_double
+        read_float
+      end
+
+      def read_bytes
+        validate(:bytes, data)
+        data.encode(Encoding::ISO_8859_1, Encoding::UTF_8)
+      end
+
+      def read_fixed(size)
+        read_bytes
+      end
+
+      def read_string
+        validate(:string, data)
+        data ? data.encode(Encoding::UTF_8) : data
+      end
+
+      def validate(data_type, data)
+        schema = PRIMITIVE_SCHEMAS[data_type]
+        unless Schema.validate(schema, data)
+          raise AvroTypeError.new(schema, data)
+        end
+      end
+    end # JsonDecoder
+
+    # Unlike DatumReader, JsonDatumReader reads the file first so that it can decode schema with the JSON data.
+    # It uses JsonDecoder to keep track of current data as it is going through the schema.
+    class JsonDatumReader < DatumReader
+      def validate(schema, data)
+        unless Schema.validate(schema, data)
+          raise AvroTypeError.new(schema, data)
+        end
+      end
+
+      def read_record_data(field, readers_field, decoder)
+        field_data = decoder.data ? decoder.data[field.name] : nil
+        field_decoder = decoder.class.new
+        field_decoder.data = field_data
```

Comment:
suggest you just pass `data` into the initialized unless you have a good reason not to do so.